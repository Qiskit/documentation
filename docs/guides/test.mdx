<Admonition type=note>
Dynamic circuit support on Qiskit Runtime is only available through early access. Premium Plan and Flex Plan users interested in participating in this early access should [contact support.](/docs/support#get-help-with-qiskit-and-ibm-quantum-platform)
</Admonition>

Dynamic circuits are powerful tools that let us measure qubits in the middle of a quantum circuit execution and then perform classical logic operations within the circuit, based on the outcome of those mid-circuit measurements.  This process is also known as _classical feedforward_. While we are still in the early days of understanding how best to take advantage of dynamic circuits, the quantum research community has already identified a number of use cases, such as the following:

* Efficient quantum state preparation, such as [W-state](https://arxiv.org/abs/2412.07969) and [MPS state](https://arxiv.org/abs/2404.16083)
* [Efficient long-range entanglement](https://journals.aps.org/prxquantum/abstract/10.1103/PRXQuantum.5.030339) between qubits on the same chip by using shallow circuits
* Efficient [sampling of IQP-like circuits](https://arxiv.org/pdf/2505.04705)

These improvements brought by dynamic circuits, however, come with trade-offs. Mid-circuit measurements and classical operations typically have longer execution time than two-qubit gates, and this increase in time might negate the benefits of reduced circuit depth. Therefore, reducing the length of mid-circuit measurement is a focus area of improvement as IBM Quantum rolls out the [new version](/announcements/product-updates/2025-03-03-new-version-dynamic-circuits) of dynamic circuits.

The [OpenQASM 3 specification](https://openqasm.com/language/classical.html#looping-and-branching) defines a number of control-flow structures, but Qiskit Runtime currently only supports the conditional `if` statement. In Qiskit SDK, this corresponds to the [if_test](/docs/api/qiskit/qiskit.circuit.QuantumCircuit#if_test) method on [QuantumCircuit.](/docs/api/qiskit/qiskit.circuit.QuantumCircuit) This method returns a [context manager](https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers) and is typically used in a `with` statement. This guide describes how to use this conditional statement.