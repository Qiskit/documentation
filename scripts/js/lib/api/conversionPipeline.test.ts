// This code is a Qiskit project.
//
// (C) Copyright IBM 2024.
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

import os from "os";
import path from "path";
import { mkdtemp, readFile } from "fs/promises";

import { globby } from "globby";
import { expect, test } from "@playwright/test";

import { runConversionPipeline } from "./conversionPipeline.js";
import { Pkg, ReleaseNotesConfig } from "./Pkg.js";

// This test uses snapshot testing (https://jestjs.io/docs/snapshot-testing#updating-snapshots). If the tests fail and the changes
// are valid, run `npm test -- --updateSnapshot`.
//
// To regenerate the input, run `tox -e docs` in https://github.com/Qiskit/qiskit_sphinx_theme,
// then copy over `stubs/`, `apidoc/`, and `objects.inv`
// to the `testdata/qiskit-sphinx-theme` folder.

// Set this to false to write the results to `docs/` in this repo.
// That is useful to let you preview how the docs render with
// `./start` and navigating to localhost:3000/api/qiskit-sphinx-theme/{fileName}
// rather than looking at raw markdown.
//
// However, don't check the output in to version control! Use
// `git clean -fd` and set this value back to `true`.
const USE_TMPDIR = true;

test("qiskit-sphinx-theme", async ({}, testInfo) => {
  // The integration test currently does not test these mechanisms:
  //   - images
  //   - release notes

  // Normally, Playwright saves the operating system name in the snapshot results.
  // Our test is OS-independent, so turn this off.
  testInfo.snapshotSuffix = "";

  let docsBaseFolder = "docs";
  let publicBaseFolder = "public";
  if (USE_TMPDIR) {
    const tmpDir = await mkdtemp(
      path.join(os.tmpdir(), "qiskit-sphinx-theme-test-"),
    );
    docsBaseFolder = path.join(tmpDir, docsBaseFolder);
    publicBaseFolder = path.join(tmpDir, publicBaseFolder);
  }

  const pkg = new Pkg({
    name: "qiskit-sphinx-theme",
    title: "Qiskit Sphinx Theme",
    githubSlug: "Qiskit/qiskit_sphinx_theme",
    version: "0.1.1",
    versionWithoutPatch: "0.1",
    type: "latest",
    releaseNotesConfig: new ReleaseNotesConfig({ enabled: false }),
  });
  const markdownFolder = pkg.outputDir(docsBaseFolder);

  await runConversionPipeline(
    "scripts/js/lib/api/testdata/qiskit-sphinx-theme",
    docsBaseFolder,
    publicBaseFolder,
    pkg,
  );

  const resultFiles = await globby([`${markdownFolder}/**`]);
  for (const file of resultFiles) {
    const contents = await readFile(file, "utf-8");
    const fileName = path.parse(file).name;
    expect(contents).toMatchSnapshot(fileName);
  }
  console.log(`Check out your project at ${markdownFolder}`);
});
