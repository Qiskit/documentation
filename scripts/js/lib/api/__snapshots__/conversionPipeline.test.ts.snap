// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qiskit-sphinx-theme: _package 1`] = `
"{
  "name": "qiskit-sphinx-theme",
  "version": "0.1.1"
}
"
`;

exports[`qiskit-sphinx-theme: _toc 1`] = `
"{
  "title": "Qiskit Sphinx Theme",
  "children": [
    {
      "title": "API index",
      "url": "/api/qiskit-sphinx-theme"
    },
    {
      "title": "api_example",
      "children": [
        {
          "title": "Module overview",
          "url": "/api/qiskit-sphinx-theme/module"
        },
        {
          "title": "Electron",
          "url": "/api/qiskit-sphinx-theme/api_example.Electron"
        },
        {
          "title": "my_function1",
          "url": "/api/qiskit-sphinx-theme/api_example.my_function1"
        }
      ]
    },
    {
      "title": "api_example.inline_classes",
      "url": "/api/qiskit-sphinx-theme/inline_classes"
    }
  ],
  "collapsed": true
}
"
`;

exports[`qiskit-sphinx-theme: api_example.Electron 1`] = `
"---
title: Electron
description: API reference for api_example.Electron
in_page_toc_min_heading_level: 1
python_api_type: class
python_api_name: api_example.Electron
---

# Electron

<Class id="api_example.Electron" isDedicatedPage={true} github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example/electron.py" signature="api_example.Electron(size=None, name=None)" modifiers="class">
  Bases: \`object\`

  A representation of an electron.

  **Examples**

  \`\`\`python
  from api_example import Electron

  ELECTRON = Electron(size="2GB", name="QuantumComputing")
  \`\`\`

  ### size

  <Attribute id="api_example.Electron.size">
    How big the Electron is.

    **Type**

    str
  </Attribute>

  ### name

  <Attribute id="api_example.Electron.name">
    What the Electron is called.
  </Attribute>

  Create an electron.

  **Parameters**

  *   **size** (*str*) – How big should this thing be?
  *   **name** (*str*) – The name we’ll call the electron. Nicknames preferred.

  **Raises**

  **ValueError** – You did something wrong

  ## Attributes

  ### CLASS\\_ATTRIBUTE

  <Attribute id="api_example.Electron.CLASS_ATTRIBUTE" attributeValue="re.compile('abc')" />

  ### charge

  <Attribute id="api_example.Electron.charge" />

  ### mass

  <Attribute id="api_example.Electron.mass">
    The mass of the electron.
  </Attribute>

  ### really\\_long\\_named\\_attribute\\_that\\_probably\\_does\\_not\\_fit\\_nicely

  <Attribute id="api_example.Electron.really_long_named_attribute_that_probably_does_not_fit_nicely">
    A bit too verbose if you ask me.
  </Attribute>

  ## Methods

  ### compute\\_momentum

  <Function id="api_example.Electron.compute_momentum" signature="compute_momentum(velocity)">
    Compute the electron’s velocity.

    **Parameters**

    **velocity** (*float*) – The electron’s velocity

    **Returns**

    The computed momentum.

    **Raises**

    **ValueError** – You did something wrong

    **Return type**

    float
  </Function>

  ### method\\_with\\_a\\_reallyyreallreallyreallyreallyreallyreallreallyreallyreallyreally\\_long\\_title

  <Function id="api_example.Electron.method_with_a_reallyyreallreallyreallyreallyreallyreallreallyreallyreallyreally_long_title" signature="method_with_a_reallyyreallreallyreallyreallyreallyreallreallyreallyreallyreally_long_title()">
    blahblahblahblahblahblahblahblahblahblahblahblahblahblahblahblahblahblah
  </Function>

  ### overloaded\\_func

  <Function id="api_example.Electron.overloaded_func" signature="overloaded_func(arg1: tuple[str, str], arg2: list[str], arg3: int, arg4: Electron) → None" extraSignatures={["overloaded_func(arg1: tuple[int, int], arg2: list[int], arg3: bool, arg4: set[Electron]) → None"]}>
    This is meant to test out [https://github.com/Qiskit/qiskit\\_sphinx\\_theme/pull/319](https://github.com/Qiskit/qiskit_sphinx_theme/pull/319).

    **Parameters**

    *   **arg1** – Tuples ftw!
    *   **arg2** – But lists are more flexy.
    *   **arg3** – Primitive values are good too.
    *   **arg4** – Recursionnnnn.
  </Function>
</Class>

"
`;

exports[`qiskit-sphinx-theme: api_example.my_function1 1`] = `
"---
title: my_function1
description: API reference for api_example.my_function1
in_page_toc_min_heading_level: 1
python_api_type: function
python_api_name: api_example.my_function1
---

<span id="api-example-my-function1" />

# api\\_example.my\\_function1

<Function id="api_example.my_function1" isDedicatedPage={true} github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example.py" signature="api_example.my_function1(input1, input2, input3=None, **kwargs)">
  A function that does awesome stuff.

  **Returns**

  Did the function work.

  **Raises**

  *   **ValueError** – If the inputs are not the correct values.
  *   **TypeError** – If the inputs are not strings.

  **Return type**

  int
</Function>

"
`;

exports[`qiskit-sphinx-theme: index 1`] = `
"# API reference

*   [API example (\`api_example\`)](module)
*   [Inline classes](inline_classes)
"
`;

exports[`qiskit-sphinx-theme: inline_classes 1`] = `
"---
title: inline_classes
description: API reference for api_example.inline_classes
in_page_toc_min_heading_level: 2
python_api_type: module
python_api_name: api_example.inline_classes
---

<span id="module-api_example.inline_classes" />

<span id="api-example-inline-classes" />

# Inline classes

This is a page to test out how we render classes and functions included inline in the page. This is common with module pages.

## Important APIs

Every time you use this program, you’ll want to create an instance of [\`api_example.inline_classes.SimpleInlineClass\`](#api_example.inline_classes.SimpleInlineClass "api_example.inline_classes.SimpleInlineClass"). It has a simple interface:

### SimpleInlineClass

<Class id="api_example.inline_classes.SimpleInlineClass" github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example/inline_classes.py" signature="api_example.inline_classes.SimpleInlineClass(arg1)" modifiers="class">
  This is a simple class that does not have any methods or attributes.

  It only has the class description and constructor. Many classes in Qiskit docs are simple like this.
</Class>

It can be useful to use free functions rather than the class:

### inline\\_function

<Function id="api_example.inline_classes.inline_function" github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example/inline_classes.py" signature="api_example.inline_classes.inline_function(input1, input2)">
  A function that does awesome stuff.

  **Returns**

  Did the function work.

  **Return type**

  int
</Function>

Sometimes, you even need to use a really complex class!

### InlineClassWithMethods

<Class id="api_example.inline_classes.InlineClassWithMethods" github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example/inline_classes.py" signature="api_example.inline_classes.InlineClassWithMethods" modifiers="class">
  Bases: \`object\`

  This class is more involved.

  Note how the methods and attributes are rendered and indented when this class is inlined on the docs page.

  #### CLASS\\_ATTRIBUTE

  <Attribute id="api_example.inline_classes.InlineClassWithMethods.CLASS_ATTRIBUTE" attributeTypeHint="str" attributeValue="'An important part of any API.'" />

  #### interest\\_rate

  <Attribute id="api_example.inline_classes.InlineClassWithMethods.interest_rate" />

  #### method1

  <Function id="api_example.inline_classes.InlineClassWithMethods.method1" signature="method1()">
    A simple method.

    **Return type**

    int
  </Function>

  #### method2

  <Function id="api_example.inline_classes.InlineClassWithMethods.method2" signature="method2(arg1, arg2, description)">
    A method with a lot of args!

    This method will use a Hamiltonian to reach quantum advantage. Hamilton: great play & the secret to quantum computing. What a polymath.

    **Parameters**

    *   **arg1** (*int | float*) – All numbers accepted.
    *   **arg2** (*list\\[*[*InlineClassWithMethods*](#api_example.inline_classes.InlineClassWithMethods "api_example.inline_classes.InlineClassWithMethods")*]*) – A list of other instances, although these will be discarded.
    *   **description** (*str*) – If your description is too boring or too cryptic, this program will crash your computer.

    **Return type**

    tuple\\[int, [*InlineClassWithMethods*](#api_example.inline_classes.InlineClassWithMethods "api_example.inline_classes.InlineClassWithMethods")]
  </Function>
</Class>

### Warning: exceptions

The above APIs might raise an exception! Be careful!

#### CustomException

<Class id="api_example.inline_classes.CustomException" github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example/inline_classes.py" signature="api_example.inline_classes.CustomException" modifiers="exception">
  See how exceptions render too.
</Class>

## Other prose

Blah blah blah.

"
`;

exports[`qiskit-sphinx-theme: module 1`] = `
"---
title: api_example
description: API reference for api_example
in_page_toc_min_heading_level: 2
python_api_type: module
python_api_name: api_example
---

<span id="api-example-api-example" />

<span id="module-api_example" />

# API example

<span id="module-api_example" />

\`api_example\`

Welcome to my super cool module!

<Admonition title="Note" type="note">
  This is an example!
</Admonition>

Testing internal references… [\`Electron.compute_momentum()\`](api_example.Electron#compute_momentum "api_example.Electron.compute_momentum").

## Contents

|                                                                                                  |                                     |
| ------------------------------------------------------------------------------------------------ | ----------------------------------- |
| [\`Electron\`](api_example.Electron "api_example.Electron")(\\[size, name])                         | A representation of an electron.    |
| [\`my_function1\`](api_example.my_function1 "api_example.my_function1")(input1, input2\\[, input3]) | A function that does awesome stuff. |

## Functions

### my\\_function2

<Function id="api_example.my_function2" github="https://github.com/Qiskit/qiskit_sphinx_theme/tree/stable/0.1/api_example.py" signature="api_example.my_function2(input1, input2, input3=None, **kwargs)">
  A function that does awesome stuff.

  **Returns**

  Did the function work.

  **Raises**

  *   **ValueError** – If the inputs are not the correct values.
  *   **TypeError** – If the inputs are not strings.

  **Return type**

  int
</Function>

"
`;
